containers/npm_publish.go

```diff
@@ -0,0 +1,86 @@
+package containers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"dagger.io/dagger"
+	"github.com/grafana/grafana-build/executil"
+)
+
+// PublishNPM publishes a npm package to the given destination.
+func PublishNPM(ctx context.Context, d *dagger.Client, pkg *dagger.File, opts *NPMOpts) (string, error) {
+	src := ExtractedArchive(d, pkg, "pkg.tgz")
+
+	version, err := GetJSONValue(ctx, d, src, "package.json", "version")
+	if err != nil {
+		return "", err
+	}
+
+	name, err := GetJSONValue(ctx, d, src, "package.json", "name")
+	if err != nil {
+		return "", err
+	}
+
+	isLatestStable, err := IsLatestGrafana(ctx, d, executil.Stable, version)
+	if err != nil {
+		return "", err
+	}
+
+	isLatestPreview, err := IsLatestGrafana(ctx, d, executil.Preview, version)
+	if err != nil {
+		return "", err
+	}
+
+	latestStable, err := GetLatestGrafanaVersion(ctx, d, executil.Stable)
+	if err != nil {
+		return "", err
+	}
+
+	latestPreview, err := GetLatestGrafanaVersion(ctx, d, executil.Preview)
+	if err != nil {
+		return "", err
+	}
+
+	isLaterThanPreview, err := IsLaterThan(ctx, d, version, latestPreview)
+	if err != nil {
+		return "", err
+	}
```

Comment:
Can you give me some context why all of this is necessary / what these functions do?

I'm a little bit worried that there's too much logic in this package that should maybe be delegated to the thing that is calling it, like for example including a `--latest` flag to publish the given npm package as the latest one, for example.

`grabpl` had a lot of similar logic and it made it really hard to follow and safely make changes to these different paths. We ended up overwriting things when we didn't mean to because a change looked benign that wasn't and so on.

Reply:
A good way we can handle this, even would be to include an argument when promoting for "latest", for example, which could apply both to `npm` and `docker`.

Reply:
I removed this crazy version parsing logic and replaced it by the `latest` flag.